Docs for the Unminify feature in Performance tool (#1784)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/evaluate-performance/unminify.md

---
title: Unminify performance profiles
description: Tutorial for unminifying the call stack in a performance profile using source maps.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.date: 03/08/2022
---
# Unminify performance profiles

> <!-- todo: review all mentions of "source mapping" in this article - what is the relation between these 3 features/technologies?

Reviewer: @captainbrosset - could you pls check how we present the benefits/UX of these 3 distinct technologies/techniques?

*  source mapping and its UX/benefits
*  unminification and its UX/benefits
*  ADO secure server (for source maps) and its UX/benefits

Reviewer 2: The thing is that these are not 3 distinct technologies. Your first bullet point is the technology that devtools uses in several places. The second point is the act of transforming code in various panels of devtools using source maps. And the third is the same thing, but also using ADO to help with security.

So I don't think it makes sense to create a table that explains what each of these are and what their benefits are. They're all linked. You can't unminify code without source maps. You can't use ADO's secure server without source maps.

Now, there's a bit of possible confusion with the word "unminify". It also sometimes refers to prettifying code in the Sources tool. When code is not nicely formatted (often all on 1 line), there's a pretty-print button at the bottom people can use to reformat the code. This feature does not use source maps behind the scenes. It parses the code to understand its syntax and produces a nicely formatted version based on this. 
I know I used to call this feature minification, and I think other people might too. Devtools now calls this "pretty printing".

What the performance tool does, with the "unminify profiles" feature is NOT pretty-printing. It uses source maps to swap production function names with original function names in perf profiles.

I think this article would benefit from removing the word "unminify", and replace this with something like "map to original function names".

2 problems with this: 
1. it's much longer
2. the devtools UI already has the word "unminify"

But, I don't think we should write a table with the UX/benefits of the 3 technologies you mentioned. 
I think we should a really good intro to source maps and how devtools uses them throughout the tools for better developer UX. And then refer to this intro in other articles, like this one.

When recording a performance profile in the **Performance** tool, the **Main** section of the profile displays detailed call stacks of JavaScript functions that were called during the recording.  

Ideally, in this call stack view, you want to see the meaningful function names that are shown in your original source code.  However, if your production code is compiled, minified, or bundled, the **Performance** tool initially displays the function names that are extracted from your transformed, _production_ code.  These production function names might be obscure and might not easily map to function names in your original source code.

Source mapping is used in DevTools to keep your client-side code readable and debuggable.  By clicking the **Unminify** button in the Performance tool, you can use source maps to create an unminified version of the performance profile you recorded.

<!-- ====================================================================== -->
## How to unminify a performance profile

To unminify a performance profile, so that some more meaningful function names are shown in addition to the minified identifiers, follow the steps below.

   **Set up source maps:**

> 1. Set up source maps, to host your source maps side-by-side with your production code.  Host your source maps as well as your production code, where the **Performance** tool can find your source maps.<!-- Can you use ado server for secure hosting of source maps?-->  For details, see [Map the processed code to your original source code, for debugging](/microsoft-edge/devtools-guide-chromium/javascript/source-maps).

Reviewer: The current format guides the user end to end with numbered steps to follow.  Your change implies using a convention of `### 3. Publish source maps` to retain numbered steps as sequential guidance for the reader.  I will try including numbers that way.

   <!-- Or, if you are just trying out this feature, you can go to a browser-based Outlook mailbox: `https://outlook.office.com`, because this web app produces minified function names on the call stack. -->

>    In the following example, the source maps are served from the same server directory as the production code:

Reviewer: @zoherghadyali - Which files in the listing .png are source map files?
Rendered view: https://review.docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/evaluate-performance/unminify?branch=pr-en-us-1784#how-to-unminify-a-performance-profile

Reviewer 2: The navigator pane does not show source maps. It either shows the files downloaded from the web server (usually minified/uglified/bundled), or the original source files (which developers are familiar with). Source maps files are the underlying technology that makes it possible to go from the former to the latter, but are not displayed in devtools (apart from the Source Map Monitor tool).

Reviewer 2: That being said, I think this line and the screenshot below it are confusing. This article talks about the performance tool, but here is a screenshot of the Sources tool, and we don't explain why, or what should we be looking at in the screenshot.

I think it might be better to remove those, and instead explain in details where to host source maps (next to "where the Performance tool can find your source maps", see my earlier comment).

   <!-- a text-based format that could replace the Sources > Page tab screenshot:
   ```
   hooks.ts
   index.js
   index.ts
   isLoki
   ```
   -->

   **Record a performance profile:**

1. In the **Performance** tool, click the **Record** button, to record a performance profile for the page you want to analyze.  For details, see [Record runtime performance](/microsoft-edge/devtools-guide-chromium/evaluate-performance/reference).

   <!-- If you are just quickly trying out the feature, you can record a performance profile while you scroll through a couple of emails and click them at your inbox at the Outlook web app. -->

   ![Recording a performance profile.](../media/evaluate-performance-performance-record-highlight.msft.png)

   A performance profile is recorded, in the **Performance** tool.

1. In the **Performance** tool, click in the main area (the call stack), and use the mouse wheel to zoom in.  Click and drag left and right.  Look at the function names in the call stack.  Select a small section of the profile and look at the call stack.  Note the timestamp range.

>    The minified call stack in the **Performance** tool shows function names as **b**, **O**, **Xt**, **bn**, **n**, **i**, or **Yn**.

Reviewer 2: It should be made clear that these function names are just examples of what minified names might look like.

   ![The minified call stack in the performance profile.](images/minified-perf-profile.msft.png)

   **Unminify the code:**

1. Click the **Unminify** button:
   ![The Unminify button in the Performance tool.](images/perf-profile-unminify-button.msft.png)

   In DevTools, a popup briefly appears, reading: **Status: creating unminfied file**.  Then the **Save As** dialog box opens.  The default filename is a pattern starting with date and time like `yymmddThhmmss`, such as: `Profile-20220307T164948-unminified.json`.  The default directory is the **Downloads** directory.

1. Select a filename and directory and then click the **Save** button.

>    The **Performance** tool creates an unminified version of the recorded performance profile.  In the unminified version of the performance profile, some of the minified function names in the call stack are prepended by more meaningful function names from your source maps, where possible.  Some of the minified production function names might not be decorated with more meaningful, source-mapped names.

Reviewer 2: ```suggestion
   The **Performance** tool creates a new version of the recorded performance profile with some function names from the flame chart replaced by more meaningful function names using your source maps. Because source maps don't always provide the necessary information for the **Performance** tool to map all function names, some of the minified production names might still appear.
```

Reviewer 2: If there is a simple way to explain why we can't map all function names, we should document it here.

An example of a performance profile flame chart, with some more meaningful function names restored:

![The unminified flame chart in the performance profile, with meaningful function names restored.](images/unminified-perf-profile.msft.png)

When recording a performance profile in the **Performance** tool, the **Main** section of the profile displays detailed flame charts of JavaScript functions that were called during the recording.  

Ideally, in this flame chart view, you want to see the meaningful function names that are shown in your original source code.  However, if your production code is compiled, minified, or bundled, the **Performance** tool initially displays the function names that are extracted from your transformed, _production_ code.  These production function names might be obscure and might not easily map to function names in your original source code.

To unminify a performance profile, so that some more meaningful function names are shown in addition to the minified identifiers, follow the steps below.

<!-- ====================================================================== -->
## Step 1 - Set up source maps

> Set up source maps, to host your source maps in addition to your production code.  Host your source maps where the **Performance** tool can find them.

Author: I don't think this really makes sense. I'd prefer the following:

```suggestion
Host your source maps side-by-side with your production code. When unminifying, the **Performance** tool expects to find your source maps along with your production code.
```

For details, see [Map the processed code to your original source code, for debugging](/microsoft-edge/devtools-guide-chromium/javascript/source-maps).

<!-- ====================================================================== -->
## Step 2 - Record a performance profile

1. In the **Performance** tool, click the **Record** button, to record a performance profile for the page you want to analyze.  For details, see [Record runtime performance](/microsoft-edge/devtools-guide-chromium/evaluate-performance/reference).

   ![Recording a performance profile.](../media/evaluate-performance-performance-record-highlight.msft.png)

   A performance profile is recorded, in the **Performance** tool.

> 1. In the **Performance** tool, click in the main area (the flame chart), and use the mouse wheel to zoom in.  Click and drag left and right.  Look at the function names in the flame chart.  Select a small section of the profile and look at the flame chart.  Note the timestamp range.

Author: Since it's not like we're giving readers a demo site, I don't think we need to be so specific about where to look at function names. It's their recording, we can just provide instructions on looking at the flame chart.

```suggestion
1. In the **Performance** tool, click in the main area (the flame chart), and use the mouse wheel to zoom in.  Click and drag left and right.  Look at the function names in the flame chart.
```

>    The minified flame chart in the **Performance** tool shows terse function names, such as **XI**, **Gi**, **d**, **o**, **an**, **c**, **co**, and **a**:

Author: ```suggestion
   The flame chart in the **Performance** tool shows minified function names, such as **XI**, **Gi**, **d**, **o**, **an**, **c**, **co**, and **a** from your production code:
```

<!-- ====================================================================== -->
## Step 4 - Load the unminified profile

1. In the **Performance** tool, click the **Load profile** (![The 'Load profile` icon button in the Performance tool.](images/load-profile-icon.png)) icon button.  Or, right-click a blank area of the **Performance** tool, and then select **Load profile**.

   ![Load Profile.](../media/evaluate-performance-performance-refreshed-disable-javascript-samples-checkbox-off-load-profile.msft.png)

1. In the **Open** dialog box, select the unminified performance profile `.json` file that you just downloaded.  If needed, for details, see [Load a recording](/microsoft-edge/devtools-guide-chromium/evaluate-performance/reference#load-a-recording) in _Performance features reference_, and then continue below.

> 1. Select the same section (timestamp range) of the profile you looked at in the first image.

Author: Like I said earlier, it's not a demo so our instructions don't need to be so perfectly step-by-step. I'd cut this entirely.
```suggestion
```

>    Some function names are now unminified.  The unminified flame chart in the **Performance** tool now shows those functions labeled with more meaningful names taken from your source maps, such as **gg (XI)**, **oh (Gi)**, **useObserver (d)**, **useQueuedForceUpdateBlock (o)**, **trackDerivedFunction (an)**, **initializeIconColorMap (c)**, **getExtraToolControls (co)**, and **isToggleEnabled (a)**.

Author: It's not really that helpful that some of these function names are still minified, just with different names. It's important to remember that there are multiple transforms between source code and production code and sourcemaps may undo one of those transformations but not all of them perfectly. I'd cut the ones that don't clearly map to unminified function names.
```suggestion
   Some function names are now unminified.  The unminified flame chart in the **Performance** tool now shows those functions labeled with more meaningful names taken from your source maps, such as **useObserver (d)**, **useQueuedForceUpdateBlock (o)**, **trackDerivedFunction (an)**, **initializeIconColorMap (c)**, **getExtraToolControls (co)**, and **isToggleEnabled (a)**.
```
